### PR TITLE
[CARBONDATA-35] generate global dict using pre-defined dict from external column file

### PR DESCRIPTION
**Pull request by Jay357089**
To support pre-defined dict from external column file, we add Key word COLUMNDICT in option of load DML to define the column name and column file path as below:

LOAD DATA inpath 'D:/inputPath/carbonTest.csv' INTO TABLE carbonTest options( 'COLUMNDICT '='name:D:/Carbon/carbonData/name.csv,sex:D:/Carbon/carbonData/sex.csv'), 
user needs to set as "column_Name: column_Path", and multiple columns and file paths can be delimited by ",'.
for complex type, user needs to set directly on primitive type. For example, for struct type(imei struct(num int, name string)), user needs set like "imei.num:columfilepath", for Array type(imei Array(string)), user can set "imei:columfilepath". for more , pls check ExternalColumnDictionaryTestCase.scala

we handle the above DML as below:
1. in LoadCube, set column name and dictionanry file path in carbonLoadModel.
2. get the column name and column file path from carbonLoadModel, test whether the column is exists, then generate new dict using the column file.
3. read the old dictionary file and compare with new dict and remove duplication, then write dict file.
4 if the column data is also provided by fact table, then drop the column data from fact table when generating dict using fact table file.

### CODE REVIEW

**Comment by Jay357089** [comment]
@QiangCai handled and have passed CI in old carbondata. pls review.

**Comment by QiangCai** [comment]
@ravipesala please check CI.

**Comment by ravipesala** [comment]
@QiangCai CI is not configured yet to this repository, so please make sure all test cases and static checks are passing in your local machine.

File: integration/spark/src/main/scala/org/carbondata/spark/rdd/CarbonGlobalDictionaryRDD.scala [hunk 1]
```diff
@@ -321,3 +327,108 @@ class CarbonGlobalDictionaryGenerateRDD(
...
+        case ex: Exception =>
+          throw new DataLoadingException(s"Error in reading pre-defined " +
+            s"dictionary file:${ex.getMessage}")
+    } finally {
+      if (csvReader != null) {
+        try {
+          csvReader.close
+        } catch {
+          case ex: Exception =>
+            throw new DataLoadingException(s"Error in closing csvReader of " +
```
Comment (QiangCai) [inline integration/spark/src/main/scala/org/carbondata/spark/rdd/CarbonGlobalDictionaryRDD.scala]: no need to throw exception.
just need to log messages.

File: integration/spark/src/main/scala/org/carbondata/spark/rdd/CarbonGlobalDictionaryRDD.scala [hunk 2]
```diff
@@ -321,3 +327,108 @@ class CarbonGlobalDictionaryGenerateRDD(
...
+    for (dimension <- dimensions) {
+      val dims = getPrimDimensionWithDict(carbonLoadModel, dimension, true)
+      primDimensionsBuffer ++= dims
+    }
+    val primDimensions = primDimensionsBuffer.toArray
+    val primDimLength = primDimensions.length
+    val result = new Array[Partition](primDimLength)
+    var primColStarIndex = 0
+    for (i <- 0 until primDimLength) {
+      result(i) = new CarbonColumnDictPatition(i, primDimensions(i))
```
Comment (QiangCai) [inline integration/spark/src/main/scala/org/carbondata/spark/rdd/CarbonGlobalDictionaryRDD.scala]: no need to define new variable "primDimensions"
please directly use variable "primDimensionsBuffer"

File: integration/spark/src/main/scala/org/carbondata/spark/rdd/CarbonGlobalDictionaryRDD.scala [hunk 3]
```diff
@@ -321,3 +327,108 @@ class CarbonGlobalDictionaryGenerateRDD(
...
+    } finally {
+      if (csvReader != null) {
+        try {
+          csvReader.close
+        } catch {
+          case ex: Exception =>
+            throw new DataLoadingException(s"Error in closing csvReader of " +
+              s"pre-defined dictionary file:${ex.getMessage}")
+        }
+      } else if (inputStream != null) {
```
Comment (QiangCai) [inline integration/spark/src/main/scala/org/carbondata/spark/rdd/CarbonGlobalDictionaryRDD.scala]: change "else if " to "if"

File: integration/spark/src/main/scala/org/carbondata/spark/util/GlobalDictionaryUtil.scala [hunk 4]
```diff
@@ -106,27 +107,36 @@ object GlobalDictionaryUtil extends Logging {
...
-        gatherDimensionByEncoding(c, encoding, excludeEncoding, dimensionsWithEncoding)
+        gatherDimensionByEncoding(carbonLoadModel, c, encoding, excludeEncoding,
+          dimensionsWithEncoding, forPreDefDict)
       }
     } else {
       if (dimension.hasEncoding(encoding) &&
           (excludeEncoding == null || !dimension.hasEncoding(excludeEncoding))) {
-        dimensionsWithEncoding += dimension
+        if ((forPreDefDict && carbonLoadModel.getPredefDictFilePath(dimension) != null) ||
+          (!forPreDefDict && carbonLoadModel.getPredefDictFilePath(dimension) == null)) {
```
Comment (QiangCai) [inline integration/spark/src/main/scala/org/carbondata/spark/util/GlobalDictionaryUtil.scala]: !forPreDefDict

File: integration/spark/src/main/scala/org/carbondata/spark/util/GlobalDictionaryUtil.scala [hunk 5]
```diff
@@ -106,27 +107,36 @@ object GlobalDictionaryUtil extends Logging {
...
 
-  def gatherDimensionByEncoding(dimension: CarbonDimension,
+  def gatherDimensionByEncoding(carbonLoadModel: CarbonLoadModel,
+      dimension: CarbonDimension,
       encoding: Encoding,
       excludeEncoding: Encoding,
-      dimensionsWithEncoding: ArrayBuffer[CarbonDimension]) {
+      dimensionsWithEncoding: ArrayBuffer[CarbonDimension],
+      forPreDefDict: Boolean) {
     if (dimension.isComplex()) {
```
Comment (QiangCai) [inline integration/spark/src/main/scala/org/carbondata/spark/util/GlobalDictionaryUtil.scala]: please remove "()"

File: integration/spark/src/main/scala/org/carbondata/spark/util/GlobalDictionaryUtil.scala [hunk 6]
```diff
@@ -470,12 +593,22 @@ object GlobalDictionaryUtil extends Logging {
       else {
         carbonLoadModel.getCsvHeader.split("" + CSVWriter.DEFAULT_SEPARATOR)
       }
-      val (requireDimension, requireColumnNames) = pruneDimensions(dimensions, headers, df.columns)
+      // generate global dict from pre defined column dict file
+      val colDictFilePath = carbonLoadModel.getColDictFilePath
+      carbonLoadModel.initPredefDictMap()
```
Comment (QiangCai) [inline integration/spark/src/main/scala/org/carbondata/spark/util/GlobalDictionaryUtil.scala]: please put inside next line

File: integration/spark/src/main/scala/org/carbondata/spark/rdd/CarbonGlobalDictionaryRDD.scala [hunk 7]
```diff
@@ -321,3 +327,108 @@ class CarbonGlobalDictionaryGenerateRDD(
...
+    val primDimension = theSplit.preDefDictDimension
+    // read the column dict data
+    val preDefDictFilePath = carbonLoadModel.getPredefDictFilePath(primDimension)
+    var csvReader: CSVReader = null
+    var inputStream: DataInputStream = null
+    var colDictData: java.util.Iterator[Array[String]] = null
+    try {
+      inputStream = FileFactory.getDataInputStream(preDefDictFilePath,
+        FileFactory.getFileType(preDefDictFilePath))
+      csvReader = new CSVReader(new InputStreamReader(inputStream, Charset.defaultCharset),
```
Comment (QiangCai) [inline integration/spark/src/main/scala/org/carbondata/spark/rdd/CarbonGlobalDictionaryRDD.scala]: please check delimiter

File: integration/spark/src/test/scala/org/carbondata/spark/util/DictionaryTestCaseUtil.scala [hunk 8]
```diff
@@ -0,0 +1,36 @@
+package org.carbondata.spark.util
```
Comment (QiangCai) [inline integration/spark/src/test/scala/org/carbondata/spark/util/DictionaryTestCaseUtil.scala]: need licence

File: integration/spark/src/test/scala/org/carbondata/spark/util/ExternalColumnDictionaryTestCase.scala [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+  test("[issue-126]Generate global dictionary from external column file") {
+    // test external column file to generate global dict
+    val carbonLoadModel = buildCarbonLoadModel(extComplexRelation, complexFilePath,
+      header, extColDictFilePath)
+    GlobalDictionaryUtil.generateGlobalDictionary(CarbonHiveContext, carbonLoadModel,
+      extComplexRelation.cubeMeta.storePath)
+
+    // check whether the dictionary is generated
+    DictionaryTestCaseUtil.checkDictionary(
+      extComplexRelation, "deviceInformationId", "10086")
```
Comment (QiangCai) [inline integration/spark/src/test/scala/org/carbondata/spark/util/ExternalColumnDictionaryTestCase.scala]: please  add incremental test case

**Comment by Jay357089** [comment]
@QiangCai handled, pls check again.

File: integration/spark/src/main/scala/org/carbondata/spark/rdd/CarbonGlobalDictionaryRDD.scala [hunk 10]
```diff
@@ -321,3 +327,108 @@ class CarbonGlobalDictionaryGenerateRDD(
...
+ * Use external column dict to generate global dictionary
+  *
+  * @param carbonLoadModel  carbon load model
+ * @param sparkContext  spark context
+ * @param table  carbon table identifier
+ * @param dimensions  carbon dimenisons having predefined dict
+ * @param hdfsLocation  carbon base store path
+ * @param dictFolderPath  path of dictionary folder
+ */
+class CarbonColumnDictGenerateRDD(carbonLoadModel: CarbonLoadModel,
```
Comment (QiangCai) [inline integration/spark/src/main/scala/org/carbondata/spark/rdd/CarbonGlobalDictionaryRDD.scala]: please try to use DictionaryLoadModel

**Comment by QiangCai** [comment]
LGTM.   but we need pass CI at first. @ravipesala

**Comment by Jay357089** [comment]
ci
ci2

@ravipesala @QiangCai  CI passed in HuaweiCarbondata. pls review and check in.

**Comment by eason-lyx** [comment]
please add testcase by using the load command with COLUMNDICT keywords:
LOAD DATA inpath 'D:/inputPath/carbonTest.csv' INTO TABLE carbonTest options( 'COLUMNDICT '='name:D:/Carbon/carbonData/name.csv,sex:D:/Carbon/carbonData/sex.csv'), 
user needs to set as "column_Name: column_Path", and multiple columns and file paths can be delimited by ",'.

**Comment by Jay357089** [comment]
@eason-lyx handled.

**Comment by Jay357089** [comment]
ci2

CI passed.

**Comment by ravipesala** [comment]
@Jay357089 @QiangCai  This PR is already raised against HuaweiBigData/carbondata, If you are planning to merge there. Please don't raise here. All commits of HuaweiBigData/carbondata will be eventually merged here.

**Comment by Jay357089** [comment]
@ravipesala  PR raised in HuaweiBigData/carbondata is just used to run CI test,  i will add an commnet there.  i wanna merge here, so pls review and comment here. thanks.

**Comment by Jay357089** [comment]
ci2
ci

**Comment by Jay357089** [comment]
@chenliang613 David has reviewed this and CI also passed. Could you please review and merge this?

**Comment by chenliang613** [comment]
Some test cases not passed, please check @Jay357089 
-----------------------------------
Tests in error:
  testloadAndGetTaskIdToSegmentsMapForDifferentSegmentLoadedConcurrently(org.carbondata.core.carbon.datastore.BlockIndexStoreTest): For input string: "carbondata/core/src/test/resources/part"
  testloadAndGetTaskIdToSegmentsMapForSameBlockLoadedConcurrently(org.carbondata.core.carbon.datastore.BlockIndexStoreTest): For input string: "carbondata/core/src/test/resources/part"

Tests run: 82, Failures: 0, Errors: 2, Skipped: 0

**Comment by Jay357089** [comment]
@chenliang613 because repository name "apache/incubator-carbondata" has a "-", in the test case, it catches substring start from "-", then it causes the failure. Somebody has corrected these test cases, after rebase, now this PR is OK. pls check again.

**Comment by chenliang613** [comment]
Still have some the below errors. i got the latest code of PR-16, test with "mvn clean -Pspark-1.6.1  package"

[INFO] carbondata ......................................... SUCCESS [  0.246 s]
[INFO] carbon-common ...................................... SUCCESS [  2.968 s]
[INFO] carbon-format ...................................... SUCCESS [  2.087 s]
[INFO] carbon-core ........................................ SUCCESS [  8.183 s]
[INFO] carbon-processing .................................. SUCCESS [  4.261 s]
[INFO] carbon-hadoop ...................................... SUCCESS [  9.201 s]
[INFO] carbon-spark ....................................... FAILURE [ 50.820 s]
[INFO] carbon-assembly .................................... SKIPPED
[INFO] carbon-examples .................................... SKIPPED
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 01:17 min
[INFO] Finished at: 2016-07-09T00:28:35+05:30
[INFO] Final Memory: 71M/589M
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.scala-tools:maven-scala-plugin:2.15.2:testCompile (testCompile) on project carbon-spark: wrap: org.apache.commons.exec.ExecuteException: Process exited with an error: 1(Exit value: 1) -> [Help 1]

**Comment by ravipesala** [comment]
LGTM. It seems @QiangCai had already reviewed this PR. @Jay357089 There are compilation issues in test case. Please fix it. We can merge it.

**Comment by ravipesala** [comment]
LGTM. @chenliang613 Please merge this PR

**Comment by Jay357089** [comment]
@chenliang613 pls review.

**Pull request closed without merging**